converged_sound_open_fw_modules_development/modules/aca_module/aca_module.cc

```diff
@@ -0,0 +1,222 @@
+
+DECLARE_LOADABLE_MODULE(AcaModule,
+                        AcaModuleFactory)
+
+
+ErrorCode::Type AcaModuleFactory::Create(
+    SystemAgentInterface &system_agent,
+    ModulePlaceholder *module_placeholder,
+    ModuleInitialSettings initial_settings
+    )
+{
+    // count of input pins formats retrieved from the ModuleInitialSettings container
+    const size_t in_pins_format_count = initial_settings.GetItem<IN_PINS_FORMAT>().GetLength();
+    // count of output pins formats retrieved from the ModuleInitialSettings container
+    const size_t out_pins_format_count = initial_settings.GetItem<OUT_PINS_FORMAT>().GetLength();
+
+    LOG_MESSAGE(    LOW, "Create(in_pins_format_count = %d, out_pins_format_count=%d)",
+                    LOG_ENTRY, in_pins_format_count, out_pins_format_count);
+
+    // check that one audio format is available for the input pin
+    if (in_pins_format_count < 1 || out_pins_format_count != 1)
+    {
+        LOG_MESSAGE(    CRITICAL, "Invalid count of input pin formats received (%d)",
+                        LOG_ENTRY, in_pins_format_count);
+        return ErrorCode::INVALID_SETTINGS;
+    }
+
+    InputPinFormat const& input_pin_format = initial_settings.GetItem<IN_PINS_FORMAT>()[0];
+    OutputPinFormat const& output_pin_format = initial_settings.GetItem<OUT_PINS_FORMAT>()[0];
+    // check that audio format retrieved for input is assigned to an existing module pin.
+    if (input_pin_format.pin_index != 0 || output_pin_format.pin_index != 0 || input_pin_format.ibs != output_pin_format.obs )
+    {
+        LOG_MESSAGE(    CRITICAL, "Retrieved audio format is associated to an invalid input pin index (%d)",
+                        LOG_ENTRY, input_pin_format.pin_index);
+        return ErrorCode::INVALID_SETTINGS;
+    }
+
+    // check that ibs can be divided by the bytes size of "samples group"
+    if ((input_pin_format.ibs * 8) % (input_pin_format.audio_fmt.bit_depth * input_pin_format.audio_fmt.number_of_channels))
+    {
+        LOG_MESSAGE(    CRITICAL, "ibs*8 shall be a multiple of samples group value: "
+                        "ibs = %d, input_bit_depth = %d.",
+                        LOG_ENTRY, input_pin_format.ibs, input_pin_format.audio_fmt.bit_depth);
+         return ErrorCode::INVALID_SETTINGS;
+    }
+
+    // Initializes AcaModule instance using the
+    // "placement syntax" of operator new
+    new (module_placeholder)AcaModule(
+        input_pin_format.audio_fmt.number_of_channels,
+        input_pin_format.audio_fmt.bit_depth,
+        input_pin_format.audio_fmt.sampling_frequency,
+        input_pin_format.ibs,
+        system_agent);
+
+    return ErrorCode::NO_ERROR;
+}
+
+// Note that purpose of the source code presented below is to demonstrate usage
+// of the ADSP System API.
+// It might not be optimized enough for efficient computation.
+uint32_t AcaModule::Process(
+    InputStreamBuffer* input_stream_buffers,
+    OutputStreamBuffer* output_stream_buffers)
```

Comment:
Where is the size of the arrays for input_stream_buffers and input_stream_buffers passed or set ?

Reply:
Per my understanding at the moment the ```Process``` function is invoked when ibs (input buffer size) is available at input. ```input_stream_buffers``` is a wrapper for array and its size is set to ```ibs``` before ```Process``` invocation.

```ibs``` (```input buffer size```), ```obs``` (```output buffer size```) are set in binmaps in ```mod_cfg```:
https://github.com/thesofproject/sof/pull/4223#discussion_r636812513

Reply:
ok, how do we deal with any ibs/obs changes at runtime (e.g a demux change). Are the binmaps values taken as default values which can then be changed at runtime ?
I also think we need to mandate that these arrays are NULL terminated.